### ISSUE TITLE
redirecting HTTP to HTTPS outside IDE  leads to  HTTP2Session - Failure generating frame HeadersFrame

### ISSUE THREAD

**Issue by hopecee** [eclipse/jetty.project#1548]
Hi.
I have already asked this question on StackOverflow but I thing here may be more appropriate.
I have been able to run the project on NetBeans with no issue but gets Exception when ran outside IDE.

```
[DEBUG] 2017-05-12 14:49:47.752 [qtp2114444063-26] HttpChannel - sendResponse info=null content=HeapByteBuffer@691a8cd[p=0,l=22316,c=32768,r=22316]={<<<<?xml version="1..../body>\n\n</html>>>>\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00...\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00} complete=true committing=true callback=Blocker@50773415{null}
[DEBUG] 2017-05-12 14:49:47.756 [qtp2114444063-26] HttpChannel - COMMIT for /public/index.xhtml on ServerHttpChannelOverHTTP2@1429fa17{r=1,c=true,a=DISPATCHED,uri=https://example.com:8443/public/index.xhtml}#13
200 null HTTP/2.0
Server: Jetty(9.3.z-SNAPSHOT)
Date: Fri, 12 May 2017 13:49:43 GMT
Set-Cookie: JSESSIONID=11oh3wbzndop7b70onjehfqpa;Path=/;Secure
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Content-Type: text/html;charset=utf-8
[DEBUG] 2017-05-12 14:49:47.759 [qtp2114444063-26] HttpChannelOverHTTP2 - HTTP2 Commit Response #13/597a3cd0:
HTTP/2.0 200 null
Server: Jetty(9.3.z-SNAPSHOT)
Date: Fri, 12 May 2017 13:49:43 GMT
Set-Cookie: JSESSIONID=11oh3wbzndop7b70onjehfqpa;Path=/;Secure
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Content-Type: text/html;charset=utf-8
[DEBUG] 2017-05-12 14:49:47.762 [qtp2114444063-26] HttpTransportOverHTTP2 - HTTP2 Response #13:
HTTP/2.0 200
Server: Jetty(9.3.z-SNAPSHOT)
Date: Fri, 12 May 2017 13:49:43 GMT
Set-Cookie: JSESSIONID=11oh3wbzndop7b70onjehfqpa;Path=/;Secure
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Content-Type: text/html;charset=utf-8
[DEBUG] 2017-05-12 15:21:43.336 [qtp2114444063-29] HttpChannel - sendResponse info=null content=HeapByteBuffer@691a8cd[p=0,l=22316,c=32768,r=22316]={<<<<?xml version="1..../body>\n\n</html>>>>\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00...\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00} complete=true committing=true callback=Blocker@673ec61a{null}
[DEBUG] 2017-05-12 15:21:43.337 [qtp2114444063-29] HttpChannel - COMMIT for /public/index.xhtml on ServerHttpChannelOverHTTP2@4b8739e6{r=1,c=true,a=DISPATCHED,uri=https://tublex.com:8443/public/index.xhtml}#13
200 null HTTP/2.0
Server: Jetty(9.3.z-SNAPSHOT)
Date: Fri, 12 May 2017 14:21:42 GMT
Set-Cookie: JSESSIONID=15fz16548bsvk1hn2zklq2epuq;Path=/;Secure
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Content-Type: text/html;charset=utf-8
[DEBUG] 2017-05-12 15:21:43.340 [qtp2114444063-29] HttpChannelOverHTTP2 - HTTP2 Commit Response #13/2c93bd95:
HTTP/2.0 200 null
Server: Jetty(9.3.z-SNAPSHOT)
Date: Fri, 12 May 2017 14:21:42 GMT
Set-Cookie: JSESSIONID=15fz16548bsvk1hn2zklq2epuq;Path=/;Secure
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Content-Type: text/html;charset=utf-8
[DEBUG] 2017-05-12 15:21:43.345 [qtp2114444063-29] HttpTransportOverHTTP2 - HTTP2 Response #13:
HTTP/2.0 200
Server: Jetty(9.3.z-SNAPSHOT)
Date: Fri, 12 May 2017 14:21:42 GMT
Set-Cookie: JSESSIONID=15fz16548bsvk1hn2zklq2epuq;Path=/;Secure
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Content-Type: text/html;charset=utf-8
[DEBUG] 2017-05-12 15:21:43.351 [qtp2114444063-29] HTTP2Session - Sending HeadersFrame@754f0f6d#13{end=false}
[DEBUG] 2017-05-12 15:21:43.353 [qtp2114444063-29] HTTP2Flusher - Appended HeadersFrame@754f0f6d#13{end=false}, frames=1
[DEBUG] 2017-05-12 15:21:43.354 [qtp2114444063-29] HTTP2Flusher - Flushing HTTP2ServerSession@2c93bd95{l:/192.168.0.3:8443 <-> r:/192.168.0.3:61576,queueSize=1,sendWindow=12582912,recvWindow=65535,streams=1,NOT_CLOSED}
[DEBUG] 2017-05-12 15:21:43.355 [qtp2114444063-29] HTTP2Flusher - Gathered for write HeadersFrame@754f0f6d#13{end=false}
[DEBUG] 2017-05-12 15:21:43.356 [qtp2114444063-29] HpackEncoder - CtxTbl[3fcfbc6d] encoding
[DEBUG] 2017-05-12 15:21:43.357 [qtp2114444063-29] HpackEncoder - encode IdxFieldS1:':status: 200' to '88'
[DEBUG] 2017-05-12 15:21:43.358 [qtp2114444063-29] HTTP2Session - Failure generating frame HeadersFrame@754f0f6d#13{end=false}
java.lang.ArrayIndexOutOfBoundsException: 1
        at org.eclipse.jetty.http.PreEncodedHttpField.putTo(PreEncodedHttpField.java:119) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.hpack.HpackEncoder.encode(HpackEncoder.java:289) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.hpack.HpackEncoder.encode(HpackEncoder.java:179) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.generator.HeadersGenerator.generateHeaders(HeadersGenerator.java:72) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.generator.HeadersGenerator.generate(HeadersGenerator.java:56) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.generator.Generator.control(Generator.java:80) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.HTTP2Session$ControlEntry.generate(HTTP2Session.java:1122) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.HTTP2Flusher.process(HTTP2Flusher.java:232) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.HTTP2Session.frame(HTTP2Session.java:668) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.HTTP2Session.frames(HTTP2Session.java:640) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.HTTP2Stream.headers(HTTP2Stream.java:87) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.server.HttpTransportOverHTTP2.commit(HttpTransportOverHTTP2.java:184) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.server.HttpTransportOverHTTP2.send(HttpTransportOverHTTP2.java:111) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.HttpChannel.sendResponse(HttpChannel.java:732) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.HttpChannel.write(HttpChannel.java:788) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:235) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:219) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.HttpOutput.close(HttpOutput.java:269) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.HttpWriter.close(HttpWriter.java:49) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.ResponseWriter.close(ResponseWriter.java:163) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.Dispatcher.commitResponse(Dispatcher.java:259) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:202) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.Dispatcher.forward(Dispatcher.java:74) [project.war:1.0-SNAPSHOT]
        at org.ocpsoft.rewrite.servlet.impl.HttpRewriteResultHandler.handleResult(HttpRewriteResultHandler.java:41) [rewrite-servlet-2.0.12.Final.jar:2.0.12.Final]
        at org.ocpsoft.rewrite.servlet.RewriteFilter.rewrite(RewriteFilter.java:268) [rewrite-servlet-2.0.12.Final.jar:2.0.12.Final]
        at org.ocpsoft.rewrite.servlet.RewriteFilter.doFilter(RewriteFilter.java:188) [rewrite-servlet-2.0.12.Final.jar:2.0.12.Final]
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759) [project.war:1.0-SNAPSHOT]
        at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:582) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1180) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:512) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:213) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.Server.handle(Server.java:534) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:320) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.HttpChannel.run(HttpChannel.java:267) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589) [project.war:1.0-SNAPSHOT]
        at java.lang.Thread.run(Unknown Source) [?:1.8.0_121]
[DEBUG] 2017-05-12 15:21:43.388 [qtp2114444063-29] HTTP2Flusher - Failing, active/queued=1/0
[DEBUG] 2017-05-12 15:21:43.389 [qtp2114444063-29] HTTP2Session - Removed remote HTTP2Stream@192e989c#13{sendWindow=12582912,recvWindow=65535,reset=false,CLOSED}
[DEBUG] 2017-05-12 15:21:43.390 [qtp2114444063-29] HttpTransportOverHTTP2 - HTTP2 Response #13 failed to commit
java.lang.ArrayIndexOutOfBoundsException: 1
        at org.eclipse.jetty.http.PreEncodedHttpField.putTo(PreEncodedHttpField.java:119) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.hpack.HpackEncoder.encode(HpackEncoder.java:289) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.hpack.HpackEncoder.encode(HpackEncoder.java:179) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.generator.HeadersGenerator.generateHeaders(HeadersGenerator.java:72) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.generator.HeadersGenerator.generate(HeadersGenerator.java:56) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.generator.Generator.control(Generator.java:80) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.HTTP2Session$ControlEntry.generate(HTTP2Session.java:1122) ~[project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.HTTP2Flusher.process(HTTP2Flusher.java:232) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.HTTP2Session.frame(HTTP2Session.java:668) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.HTTP2Session.frames(HTTP2Session.java:640) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.HTTP2Stream.headers(HTTP2Stream.java:87) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.server.HttpTransportOverHTTP2.commit(HttpTransportOverHTTP2.java:184) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.http2.server.HttpTransportOverHTTP2.send(HttpTransportOverHTTP2.java:111) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.HttpChannel.sendResponse(HttpChannel.java:732) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.HttpChannel.write(HttpChannel.java:788) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:235) [project.war:1.0-SNAPSHOT]
        at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:219) [project.war:1.0-SNAPSHOT]
```

Why HTTP2Session - Failure generating frame HeadersFrame?

**Comment by joakime** [comment]
Unrelated to your issue, but know that a big live-war like you have defined has many issues.

Don't put your jetty classes in the root of the war, that's a security problem.
When you shade/merge/copy the classes, be sure you maintain/merge the various manifest information, your log shows that you are not doing that.
Be sure you copy the resources from the jetty jars too.

This has all been reproduced in an example project you can find at 
https://github.com/jetty-project/embedded-jetty-live-war

Don't use HTTP/2 from your live war, as the alpn-boot.jar requirement makes it not viable as a simple live war (the user of your live war will have extra steps to start the war), and the alpn-boot.jar is highly JVM dependent and often needs updates before it will work on a new JVM (meaning the users will have strange / oddball errors if they upgrade to a newer JVM before alpn-boot.jar exists for it, or they don't understand the hard JVM version to alpn-boot version requirement)

**Comment by sbordet** [comment]
Your issue is due to the fact that you don't have the required jars in the classpath.
In particular:

```
java.lang.ArrayIndexOutOfBoundsException: 1
        at org.eclipse.jetty.http.PreEncodedHttpField.putTo(PreEncodedHttpField.java:119) ~[project.war:1.0-SNAPSHOT]
```

happens when `http2-hpack-<version>.jar is not in the server classpath so that `HpackFieldPreEncoder` cannot be found via `ServiceLoader`.

**Comment by joakime** [comment]
Ah, then it is a bad merge of metadata.
@hopecee fix your dependencies + shading + metadata merging and you'll be fine.

**Comment by hopecee** [comment]
Thank you very much, joakime and sbordet. I found out that unpacking the Jetty server jars with `maven-dependency-plugin` was not merging the `META-INF.services` of `jetty-http-<version>.jar`  `http2-hpack-<version>.jar`. I have merged the two and it is now working, as the `ServiceLoader` is now loading the `HpackFieldPreEncoder` . 

Taking advice from joakime, I have removed my jetty classes from the root of the war to `WEB-INF.jetty-server`, and using `JettyBootstrap` at the root to start the server, as done in the live-war.

The issue of HTTP/2 from your live war, this project will be used by us, so the  requirement of alpn-boot.jar may not be much problem.

If there is any other issue regarding this setup I will be pleased to know.

Thanks again.
